### PR TITLE
Add custom heading adapter

### LINKED ISSUE #267: Provide custom heading adapter
The built-in renderer for headings is totally 100% fine for most use cases, but this is something that I usually prefer to customize if possible. To that end, comrak should enable users to provide a custom heading adapter akin to the `SyntaxHighlighterAdapter`.

### PR DESCRIPTION
**Pull request by lucperkins**
Fixes #267.

This PR adds a custom heading adapter akin to `SyntaxHighlighterAdapter` that can be used as the basis for plugins. Let me know if this is something you'd like to add and if you have any feedback on the API. To give an example, I have a project where I'd like to add some Tailwind/AlpineJS goodies directly into my headings, like this:

```html
<h2 id="some-heading" x-data="{ showAnchor: false }" @mouseover="showAnchor = true" @mouseout="showAnchor = false">
  Some heading
  <a href="#some-heading" class="ml-2" x-show="showAnchor">#</a>
</h2>
```

One of many possible use cases, of course, but always nice to have a lever to pull 😄

Come to think of it, enabling custom headings would be a way to avoid issues like #250 and #261 without hard-coding API changes.

### CODE REVIEW

File: src/tests.rs [hunk 1]
```diff
@@ -1398,6 +1398,7 @@ fn exercise_full_api<'a>() {
     let _ = ::ComrakPlugins {
         render: ::ComrakRenderPlugins {
             codefence_syntax_highlighter: Some(&syntax_highlighter_adapter),
+            heading_adapter: None,
```
Comment (kivikakk) [inline src/tests.rs:1401]: Can we add a silly little mock adapter here to similarly track calling its API?

File: src/html.rs [hunk 2]
```diff
@@ -457,22 +459,40 @@ impl<'o> HtmlFormatter<'o> {
...
+                            let heading = HeadingMeta {
+                                level,
+                                content: String::from(content),
+                            };
+
+                            let rendered = adapter.render(&heading);
+                            write!(self.output, "{}", rendered)?;
+                        }
+                        None => {
+                            write!(self.output, "<h{}>", nch.level)?;
```
Comment (kivikakk) [inline src/html.rs:479]:
```suggestion
                            self.cr()?;
                            write!(self.output, "<h{}>", nch.level)?;
```

File: src/html.rs [hunk 3]
```diff
@@ -457,22 +459,40 @@ impl<'o> HtmlFormatter<'o> {
...
+                            let level = nch.level;
+                            let mut text_content = Vec::with_capacity(20);
+                            self.collect_text(node, &mut text_content);
+
+                            let content = &String::from_utf8(text_content).unwrap();
+
+                            let heading = HeadingMeta {
+                                level,
+                                content: String::from(content),
+                            };
```
Comment (kivikakk) [inline src/html.rs:473]:
```suggestion
                            let content = String::from_utf8(text_content).unwrap();

                            let heading = HeadingMeta {
                                level,
                                content,
                            };
```

**Review by kivikakk — APPROVED** [review]
Outstanding. It occurs to me we probably want to let the heading adapter also take care of rendering the close tag? Otherwise they're restricted to only leaving open the exact `<h1>`…`<h6>` corresponding to the `level` given, since that's what we then close on the other side. This rules out uses where users may wish to transform the levels (e.g. render `#` as an `<h4>`), use a different tag entirely, nest the heading content in multiple tags, etc.

**Comment by kivikakk** [comment]
Let me know what you think of my suggestion; happy to merge whenever you're happy with what you've got, just mark as ready.

**Comment by lucperkins** [comment]
@kivikakk So, I chose not to let the adapter render the closing tag because I'm not quite sure how to pass in the AST content of the heading to the adapter. Take this heading as an example:

```markdown
## Here is some `code` and some **bold** and some *italics* and even ~strikethrough~
```

In this case, you wouldn't want just the string "Here is some code and some bold and..." You'd want those nodes, too. If there's a good way to provide that, then we could indeed provide *full* control over the rendering.

**Comment by kivikakk** [comment]
> As you can see, the adapter trait now has an `enter` method and an `exit` method. This preserves the content of the heading while providing total control over everything that wraps that content. I think it still might be better to just provide total control and pass in the level and the AST content and expect a string back. But I'm not quite sure how to do that.

I think this is perfect for now — giving the adapter the entire sub-AST to render could be problematic, because presumably it'd need some way to call back into the renderer (with appropriate state) if it just wanted to do business as usual. If a need for more complicated transforms comes up in the future, we can deal with it then.

**Comment by kivikakk** [comment]
I'm happy to merge this if you are. 👍

**Review by kivikakk — APPROVED** [review]
Awesome. Thank you so much!! :blush::heart:

**Pull request merged**